From 780ada139a938cb04524824f1409ef54f4c2b14c Mon Sep 17 00:00:00 2001
From: Gabe Lapingcao <mgultra2018@gmail.com>
Date: Sat, 12 Sep 2026 09:31:15 -0700
Subject: [PATCH 14/14] [tests/gtest]: refactor link-state data struct send
 logic

Refactors both concrete implementations of the templatized
`SendUpdateMessage` function within `LinkStateTest` to only delete links
or prefixes instead of deleting nodes advertising such link-state
objects. This fixes coverage on `bgp_ls_withdraw_link` and
`bgp_ls_withdraw_prefix` located in bgpd/bgp_ls_ted.c, which was
previously left untouched in coverage tests.

Signed-off-by: Gabe Lapingcao <mgultra2018@gmail.com>
---
 .../include/test_bgp_ls_linkstate_update.h    | 44 ++++++++++++-------
 1 file changed, 28 insertions(+), 16 deletions(-)

diff --git a/tests/gtest/include/test_bgp_ls_linkstate_update.h b/tests/gtest/include/test_bgp_ls_linkstate_update.h
index 5ab1428c79..5a976ee8a1 100644
--- a/tests/gtest/include/test_bgp_ls_linkstate_update.h
+++ b/tests/gtest/include/test_bgp_ls_linkstate_update.h
@@ -307,20 +307,27 @@ class LinkStateTest : public testing::TestWithParam<TestCase<T>> {
 
     AttributesToFrr(apiMessage.data, adv_node_id, attr);
 
-    struct ls_node* remote_node =
-        ls_node_new(remote_node_id, in_addr{}, attr->standard.remote6);
-    SendNodeMessage(*remote_node, apiMessage.event);
+    // send messages to the BGP instance
 
-    struct ls_node* adv_node =
-        ls_node_new(adv_node_id, in_addr{}, attr->standard.local6);
-    SendNodeMessage(*adv_node, apiMessage.event);
+    if (apiMessage.event == BEvent::DELETE) {
+      SendAttributesMessage(*attr, remote_node_id, apiMessage.event);
+      SendAttributesMessage(*attr, remote_node_id, apiMessage.event, true);
+    } else {
+      struct ls_node* remote_node =
+          ls_node_new(remote_node_id, in_addr{}, attr->standard.remote6);
+      SendNodeMessage(*remote_node, apiMessage.event);
+
+      struct ls_node* adv_node =
+          ls_node_new(adv_node_id, in_addr{}, attr->standard.local6);
+      SendNodeMessage(*adv_node, apiMessage.event);
 
-    SendAttributesMessage(*attr, remote_node_id, apiMessage.event);  // forward
-    SendAttributesMessage(*attr, remote_node_id, apiMessage.event,
-                          true);  // reverse
+      SendAttributesMessage(*attr, remote_node_id, apiMessage.event);
+      SendAttributesMessage(*attr, remote_node_id, apiMessage.event, true);
+
+      ls_node_del(adv_node);
+      ls_node_del(remote_node);
+    }
 
-    ls_node_del(adv_node);
-    ls_node_del(remote_node);
     ls_attributes_del(attr);
   }
 
@@ -344,13 +351,18 @@ class LinkStateTest : public testing::TestWithParam<TestCase<T>> {
 
     PrefixToFrr(apiMessage.data, adv_node_id, pref);
 
-    struct ls_node* adv_node =
-        ls_node_new(adv_node_id, in_addr{}, pref->pref.u.prefix6);
-    SendNodeMessage(*adv_node, apiMessage.event);
+    if (apiMessage.event == BEvent::DELETE) {
+      SendPrefixMessage(*pref, apiMessage.event);
+    } else {
+      struct ls_node* adv_node =
+          ls_node_new(adv_node_id, in_addr{}, pref->pref.u.prefix6);
+      SendNodeMessage(*adv_node, apiMessage.event);
+
+      SendPrefixMessage(*pref, apiMessage.event);
 
-    SendPrefixMessage(*pref, apiMessage.event);
+      ls_node_del(adv_node);
+    }
 
-    ls_node_del(adv_node);
     ls_prefix_del(pref);
   }
 
-- 
2.34.1

